File: IAIso-v5.0/core/iaiso-java/iaiso-core/src/main/java/io/iaiso/core/PressureEngine.java
package io.iaiso.core;

import io.iaiso.audit.Event;
import io.iaiso.audit.NullSink;
import io.iaiso.audit.Sink;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.locks.ReentrantLock;

/**
 * The IAIso pressure engine. Thread-safe; {@code step()} may be called
 * from multiple threads, though semantically each execution should be
 * driven by a single thread.
 *
 * <p>See {@code spec/pressure/README.md} for normative semantics.
 */
public final class PressureEngine {

    private final PressureConfig cfg;
    private final String executionId;
    private final Sink audit;
    private final Clock clock;
    private final Clock timestampClock;
    private final ReentrantLock lock = new ReentrantLock();

    // mutable state guarded by `lock`
    private double pressure;
    private long step;
    private Lifecycle lifecycle;
    private double lastDelta;
    private double lastStepAt;

    public PressureEngine(PressureConfig cfg, EngineOptions opts) {
        cfg.validate();
        this.cfg = cfg;
        this.executionId = opts.getExecutionId();
        this.audit = opts.getAuditSink() != null ? opts.getAuditSink() : NullSink.INSTANCE;
        this.clock = opts.getClock() != null ? opts.getClock() : Clock.wallclock();
        this.timestampClock = opts.getTimestampClock() != null
            ? opts.getTimestampClock()
            : this.clock;
        this.pressure = 0.0;
        this.step = 0L;
        this.lifecycle = Lifecycle.INIT;
        this.lastDelta = 0.0;
        this.lastStepAt = this.clock.now();

        Map<String, Object> data = new LinkedHashMap<>();
        data.put("pressure", 0.0);
        emit("engine.init", data);
    }

    public PressureConfig getConfig() { return cfg; }
    public String getExecutionId() { return executionId; }

    public double getPressure() {
        lock.lock();
        try { return pressure; } finally { lock.unlock(); }
    }

    public Lifecycle getLifecycle() {
        lock.lock();
        try { return lifecycle; } finally { lock.unlock(); }
    }

    public PressureSnapshot snapshot() {
        lock.lock();
        try {
            return new PressureSnapshot(pressure, step, lifecycle, lastDelta, lastStepAt);
        } finally {
            lock.unlock();
        }
    }

    /** Account for a unit of work; advance the engine. */
    public StepOutcome step(StepInput work) {
        // Fast path: locked
        lock.lock();
        try {
            if (lifecycle == Lifecycle.LOCKED) {
                Map<String, Object> data = new LinkedHashMap<>();
                data.put("reason", "locked");
                data.put("requested_tokens", work.getTokens());
                data.put("requested_tools", work.getToolCalls());
                emit("engine.step.rejected", data);
                return StepOutcome.LOCKED;
            }

            double now = clock.now();
            double elapsed = Math.max(0.0, now - lastStepAt);

            double delta = (work.getTokens() / 1000.0) * cfg.getTokenCoefficient()
                + work.getToolCalls() * cfg.getToolCoefficient()
                + work.getDepth() * cfg.getDepthCoefficient();
            double decay = cfg.getDissipationPerStep()
                + elapsed * cfg.getDissipationPerSecond();

            double newPressure = clamp01(pressure + delta - decay);
            this.pressure = newPressure;
            this.step += 1;
            this.lastDelta = delta - decay;
            this.lastStepAt = now;
            this.lifecycle = Lifecycle.RUNNING;

            Map<String, Object> stepData = new LinkedHashMap<>();
            stepData.put("step", step);
            stepData.put("pressure", pressure);
            stepData.put("delta", delta);
            stepData.put("decay", decay);
            stepData.put("tokens", work.getTokens());
            stepData.put("tool_calls", work.getToolCalls());
            stepData.put("depth", work.getDepth());
            stepData.put("tag", work.getTag());

            double pressureNow = pressure;
            double releaseThr = cfg.getReleaseThreshold();
            double escThr = cfg.getEscalationThreshold();
            boolean postReleaseLock = cfg.isPostReleaseLock();

            emit("engine.step", stepData);

            if (pressureNow >= releaseThr) {
                Map<String, Object> rd = new LinkedHashMap<>();
                rd.put("pressure", pressureNow);
                rd.put("threshold", releaseThr);
                emit("engine.release", rd);
                this.pressure = 0.0;
                if (postReleaseLock) {
                    this.lifecycle = Lifecycle.LOCKED;
                    Map<String, Object> ld = new LinkedHashMap<>();
                    ld.put("reason", "post_release_lock");
                    emit("engine.locked", ld);
                } else {
                    this.lifecycle = Lifecycle.RUNNING;
                }
                return StepOutcome.RELEASED;
            }
            if (pressureNow >= escThr) {
                this.lifecycle = Lifecycle.ESCALATED;
                Map<String, Object> ed = new LinkedHashMap<>();
                ed.put("pressure", pressureNow);
                ed.put("threshold", escThr);
                emit("engine.escalation", ed);
                return StepOutcome.ESCALATED;
            }
            return StepOutcome.OK;
        } finally {
            lock.unlock();
        }
    }

    /** Reset the engine. Emits {@code engine.reset}. */
    public PressureSnapshot reset() {
        lock.lock();
        try {
            this.pressure = 0.0;
            this.step = 0L;
            this.lastDelta = 0.0;
            this.lastStepAt = clock.now();
            this.lifecycle = Lifecycle.INIT;
            Map<String, Object> data = new LinkedHashMap<>();
            data.put("pressure", 0.0);
            emit("engine.reset", data);
            return new PressureSnapshot(pressure, step, lifecycle, lastDelta, lastStepAt);
        } finally {
            lock.unlock();
        }
    }

    private void emit(String kind, Map<String, Object> data) {
        audit.emit(new Event(executionId, kind, timestampClock.now(), data));
    }

    private static double clamp01(double v) {
        if (v < 0.0) return 0.0;
        if (v > 1.0) return 1.0;
        return v;
    }
}

File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Core/PressureEngine.cs
using System.Collections.Generic;
using Iaiso.Audit;

namespace Iaiso.Core;

/// <summary>
/// The IAIso pressure engine. Thread-safe; <c>Step()</c> may be called
/// from multiple threads, though semantically each execution should be
/// driven by a single thread. See <c>spec/pressure/README.md</c> for
/// normative semantics.
/// </summary>
public sealed class PressureEngine
{
    private readonly PressureConfig _cfg;
    private readonly string _executionId;
    private readonly ISink _audit;
    private readonly IClock _clock;
    private readonly IClock _timestampClock;
    private readonly object _lock = new();

    // mutable state guarded by `_lock`
    private double _pressure;
    private long _step;
    private Lifecycle _lifecycle;
    private double _lastDelta;
    private double _lastStepAt;

    public PressureEngine(PressureConfig cfg, EngineOptions opts)
    {
        cfg.Validate();
        _cfg = cfg;
        _executionId = opts.ExecutionId;
        _audit = opts.AuditSink ?? NullSink.Instance;
        _clock = opts.Clock ?? WallClock.Instance;
        _timestampClock = opts.TimestampClock ?? _clock;
        _pressure = 0.0;
        _step = 0;
        _lifecycle = Lifecycle.Init;
        _lastDelta = 0.0;
        _lastStepAt = _clock.Now();

        var data = new SortedDictionary<string, object?> { ["pressure"] = 0.0 };
        Emit("engine.init", data);
    }

    public PressureConfig Config => _cfg;
    public string ExecutionId => _executionId;

    public double Pressure
    {
        get { lock (_lock) { return _pressure; } }
    }

    public Lifecycle Lifecycle
    {
        get { lock (_lock) { return _lifecycle; } }
    }

    public PressureSnapshot Snapshot()
    {
        lock (_lock)
        {
            return new PressureSnapshot(_pressure, _step, _lifecycle, _lastDelta, _lastStepAt);
        }
    }

    /// <summary>Account for a unit of work; advance the engine.</summary>
    public StepOutcome Step(StepInput work)
    {
        lock (_lock)
        {
            if (_lifecycle == Lifecycle.Locked)
            {
                var rejected = new SortedDictionary<string, object?>
                {
                    ["reason"] = "locked",
                    ["requested_tokens"] = work.Tokens,
                    ["requested_tools"] = work.ToolCalls,
                };
                Emit("engine.step.rejected", rejected);
                return StepOutcome.Locked;
            }

            double now = _clock.Now();
            double elapsed = System.Math.Max(0.0, now - _lastStepAt);

            double delta = (work.Tokens / 1000.0) * _cfg.TokenCoefficient
                + work.ToolCalls * _cfg.ToolCoefficient
                + work.Depth * _cfg.DepthCoefficient;
            double decay = _cfg.DissipationPerStep + elapsed * _cfg.DissipationPerSecond;

            double newPressure = Clamp01(_pressure + delta - decay);
            _pressure = newPressure;
            _step += 1;
            _lastDelta = delta - decay;
            _lastStepAt = now;
            _lifecycle = Lifecycle.Running;

            var stepData = new SortedDictionary<string, object?>
            {
                ["step"] = _step,
                ["pressure"] = _pressure,
                ["delta"] = delta,
                ["decay"] = decay,
                ["tokens"] = work.Tokens,
                ["tool_calls"] = work.ToolCalls,
                ["depth"] = work.Depth,
                ["tag"] = work.Tag,
            };

            double pressureNow = _pressure;
            double releaseThr = _cfg.ReleaseThreshold;
            double escThr = _cfg.EscalationThreshold;
            bool postReleaseLock = _cfg.PostReleaseLock;

            Emit("engine.step", stepData);

            if (pressureNow >= releaseThr)
            {
                var rd = new SortedDictionary<string, object?>
                {
                    ["pressure"] = pressureNow,
                    ["threshold"] = releaseThr,
                };
                Emit("engine.release", rd);
                _pressure = 0.0;
                if (postReleaseLock)
                {
                    _lifecycle = Lifecycle.Locked;
                    var ld = new SortedDictionary<string, object?>
                    {
                        ["reason"] = "post_release_lock",
                    };
                    Emit("engine.locked", ld);
                }
                else
                {
                    _lifecycle = Lifecycle.Running;
                }
                return StepOutcome.Released;
            }
            if (pressureNow >= escThr)
            {
                _lifecycle = Lifecycle.Escalated;
                var ed = new SortedDictionary<string, object?>
                {
                    ["pressure"] = pressureNow,
                    ["threshold"] = escThr,
                };
                Emit("engine.escalation", ed);
                return StepOutcome.Escalated;
            }
            return StepOutcome.Ok;
        }
    }

    /// <summary>Reset the engine. Emits <c>engine.reset</c>.</summary>
    public PressureSnapshot Reset()
    {
        lock (_lock)
        {
            _pressure = 0.0;
            _step = 0;
            _lastDelta = 0.0;
            _lastStepAt = _clock.Now();
            _lifecycle = Lifecycle.Init;
            var data = new SortedDictionary<string, object?> { ["pressure"] = 0.0 };
            Emit("engine.reset", data);
            return new PressureSnapshot(_pressure, _step, _lifecycle, _lastDelta, _lastStepAt);
        }
    }

    private void Emit(string kind, IDictionary<string, object?> data)
    {
        _audit.Emit(new Event(_executionId, kind, _timestampClock.Now(),
            new SortedDictionary<string, object?>(data)));
    }

    private static double Clamp01(double v)
    {
        if (v < 0.0) return 0.0;
        if (v > 1.0) return 1.0;
        return v;
    }
}
